{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b509814a",
   "metadata": {},
   "source": [
    "# Working with modern sequence formats\n",
    "\n",
    "In this notebook, we'll use the Humans 1,000 genomes project data to explore\n",
    "the usage of FASTQ files with their quality score for each base.\n",
    "\n",
    "One of the main challenges of using NGS data is the raw size of the data.\n",
    "Due to it's large nature, we must be aware of ways to handle disk space and\n",
    "backup policies."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c38cc385",
   "metadata": {},
   "outputs": [],
   "source": [
    "!rm -f SRR003265.filt.fastq.gz 2>/dev/null\n",
    "!wget -nd ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/phase3/data/NA18489/sequence_read/SRR003265.filt.fastq.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cfceed05",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "from Bio import SeqIO\n",
    "recs = SeqIO.parse(gzip.open('SRR003265.filt.fastq.gz', 'rt', encoding='utf-8'), 'fastq')\n",
    "rec = next(recs)\n",
    "print(rec.id, rec.description, rec.seq)\n",
    "print(rec.letter_annotations)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c63b6a6",
   "metadata": {},
   "source": [
    "note that here `recs` holds an iterator with the contents of the FASTQ file. This means\n",
    "that if we were to convert this iterator to a list, it might eat our whole RAM. **The \n",
    "safest way to manipulate data in a FASTQ file is to either do all manipulations in a \n",
    "single iteration or open and close the file multiple times over.**\n",
    "\n",
    "### taking a look at the distribution of nucleotide reads:\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "379c3171",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "recs = SeqIO.parse(gzip.open('SRR003265.filt.fastq.gz', 'rt', encoding='utf-8'), 'fastq')\n",
    "cnt = defaultdict(int)\n",
    "for rec in recs:\n",
    "  for letter in rec.seq:\n",
    "    cnt[letter] += 1\n",
    "tot = sum(cnt.values())\n",
    "for letter, count in cnt.items():\n",
    "  print(f\"{letter}: {(100.*count/tot):.2f} {count}\")"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "cell_metadata_filter": "-all",
   "formats": "ipynb,md",
   "main_language": "python"
  },
  "kernelspec": {
   "display_name": "bio_genbank",
   "language": "python",
   "name": "bio_genbank"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
