{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b8836377",
   "metadata": {},
   "source": [
    "# Acessing NCBI database for gene retrieval"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5149bc99",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez, SeqIO\n",
    "Entrez.email = \"guilhormo.47@gmail.com\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1e5db5f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "handle = Entrez.esearch(db=\"nucleotide\", term='CRT[Gene Name] AND \"Plasmodium falciparum\"[Organism]')\n",
    "rec_list = Entrez.read(handle)\n",
    "if int(rec_list['RetMax']) < int(rec_list['Count']):\n",
    "    handle = Entrez.esearch(db=\"nucleotide\", term='CRT[Gene Name] AND \"Plasmodium falciparum\"[Organism]',\n",
    "                            retmax=rec_list['Count'])\n",
    "    rec_list = Entrez.read(handle)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "645900cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "id_list = rec_list['IdList']\n",
    "hdl = Entrez.efetch(db='nucleotide', id=id_list, rettype='gb', retmax=rec_list['Count'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1ef93e81",
   "metadata": {},
   "outputs": [],
   "source": [
    "recs = list(SeqIO.parse(hdl, 'gb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "da74bfc0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KM288867\n",
      "Plasmodium falciparum clone PF3D7_0709000 chloroquine resistance transporter (CRT) gene, complete cds\n"
     ]
    }
   ],
   "source": [
    "for rec in recs:\n",
    "    if rec.name == 'KM288867':\n",
    "        break\n",
    "print(rec.name)\n",
    "print(rec.description)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99b9aaa4",
   "metadata": {},
   "outputs": [],
   "source": [
    "for feature in rec.features:\n",
    "    if feature.type == 'gene':\n",
    "        print(feature.qualifiers['gene'])\n",
    "    elif feature.type == 'exon':\n",
    "        loc = feature.location\n",
    "        print('Exon', loc.start, loc.end, loc.strand)\n",
    "    else:\n",
    "        print('not processed:\\n%s' % feature)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43b4d147",
   "metadata": {},
   "outputs": [],
   "source": [
    "for name, value in rec.annotations.items():\n",
    "  print(f'{name}={value}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "651535ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(rec.seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4b294b9",
   "metadata": {},
   "source": [
    "## Performing basic sequence analysis\n",
    "\n",
    "We will be using the human **lactase** gene as an example. We will use the\n",
    "same methods as above for retrieving it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4de4173e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez, SeqIO, SeqRecord\n",
    "Entrez.email=\"guilhormo.47@gmail.com\"\n",
    "hdl = Entrez.efetch(db='nucleotide', id=['NM_002299'], rettype='gb') # lactase gene\n",
    "gb_rec = SeqIO.read(hdl, 'gb')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "675f3f9a",
   "metadata": {},
   "source": [
    "Now that we have the record, we'll extract the gene sequence, even though it\n",
    "contains more than that."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "334ab4cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "for feature in gb_rec.features:\n",
    "  if feature.type == 'CDS':\n",
    "    location = feature.location # Note translation existing\n",
    "cds = SeqRecord.SeqRecord(gb_rec.seq[location.start:location.end], 'NM_002299', description='LCT CDS only')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e91c185",
   "metadata": {},
   "source": [
    "Now our sequence is available in the Biopython sequence record.\n",
    "\n",
    "We'll now proceed by saving the sequence to a FASTA file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e6bcb94",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "with open('example.fasta', 'w') as w_hdl:\n",
    "  SeqIO.write([cds], w_hdl, 'fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec86a0dc",
   "metadata": {},
   "source": [
    "The Seq.IO write method takes a list of sequences to write. In our case it is just one,\n",
    "but we should be careful - if we plan to write lots of sequences, an iterator\n",
    "will allocate less memory from our computer.\n",
    "\n",
    "In most situations we will have the sequence loaded on the disk and will\n",
    "want to read it. For that, we have the `SeqIO.read()` method:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f15acbbc",
   "metadata": {},
   "outputs": [],
   "source": [
    "recs = SeqIO.parse('example.fasta', 'fasta')\n",
    "for rec in recs:\n",
    "  seq = rec.seq\n",
    "  print(rec.description)\n",
    "  print(seq[:10])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71cb9065",
   "metadata": {},
   "source": [
    "As we now have pure DNA, we can transcribe it:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cab269b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna = seq.transcribe()\n",
    "print(rna)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee0c14b1",
   "metadata": {},
   "source": [
    "And then, we can translate the gene into a protein:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54baca52",
   "metadata": {},
   "outputs": [],
   "source": [
    "prot = seq.translate()\n",
    "print(prot)"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "formats": "ipynb,md"
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
