{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b8836377",
   "metadata": {},
   "source": [
    "# Accessing NCBI database for data retrieval\n",
    "\n",
    "NCBI stands for National Center for Biotechnology Information.\n",
    "They possess many databases for biological data, as well as journals\n",
    "and articles of a wide variety of topics. It is common for sequencing\n",
    "data to be uploaded to the website, as well as other data from \n",
    "biological experiments.\n",
    "\n",
    "Biopython has a built-in object to access NCBI and retrieve data from \n",
    "their databases. That is the `Entrez` object, which we'll explore below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5149bc99",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "from Bio import Entrez, SeqIO\n",
    "Entrez.email = \"guilhormo.47@gmail.com\" # setting an email is good practice"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "92c73090",
   "metadata": {},
   "source": [
    "## `EInfo`: obtaining information about Entrez databases:\n",
    "\n",
    "EInfo provides interesting information about each of Entrez' databases,\n",
    "such as field index term counts, last update, available links, and \n",
    "databases' names. We'll obtain here all databases' names:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "c9fe610f",
   "metadata": {},
   "outputs": [],
   "source": [
    "stream = Entrez.einfo()\n",
    "result = stream.read()\n",
    "stream.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e30978e",
   "metadata": {},
   "source": [
    "The `result` variable now contains a list of databases in XML format (\n",
    "as is standard for Entrez search return types):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "0c88a28c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "b'<?xml version=\"1.0\" encoding=\"UTF-8\" ?>\\n<!DOCTYPE eInfoResult PUBLIC \"-//NLM//DTD einfo 20190110//EN\" \"https://eutils.ncbi.nlm.nih.gov/eutils/dtd/20190110/einfo.dtd\">\\n<eInfoResult>\\n<DbList>\\n\\n\\t<DbName>pubmed</DbName>\\n\\t<DbName>protein</DbName>\\n\\t<DbName>nuccore</DbName>\\n\\t<DbName>ipg</DbName>\\n\\t<DbName>nucleotide</DbName>\\n\\t<DbName>structure</DbName>\\n\\t<DbName>genome</DbName>\\n\\t<DbName>annotinfo</DbName>\\n\\t<DbName>assembly</DbName>\\n\\t<DbName>bioproject</DbName>\\n\\t<DbName>biosample</DbName>\\n\\t<DbName>blastdbinfo</DbName>\\n\\t<DbName>books</DbName>\\n\\t<DbName>cdd</DbName>\\n\\t<DbName>clinvar</DbName>\\n\\t<DbName>gap</DbName>\\n\\t<DbName>gapplus</DbName>\\n\\t<DbName>grasp</DbName>\\n\\t<DbName>dbvar</DbName>\\n\\t<DbName>gene</DbName>\\n\\t<DbName>gds</DbName>\\n\\t<DbName>geoprofiles</DbName>\\n\\t<DbName>medgen</DbName>\\n\\t<DbName>mesh</DbName>\\n\\t<DbName>nlmcatalog</DbName>\\n\\t<DbName>omim</DbName>\\n\\t<DbName>orgtrack</DbName>\\n\\t<DbName>pmc</DbName>\\n\\t<DbName>proteinclusters</DbName>\\n\\t<DbName>pcassay</DbName>\\n\\t<DbName>protfam</DbName>\\n\\t<DbName>pccompound</DbName>\\n\\t<DbName>pcsubstance</DbName>\\n\\t<DbName>seqannot</DbName>\\n\\t<DbName>snp</DbName>\\n\\t<DbName>sra</DbName>\\n\\t<DbName>taxonomy</DbName>\\n\\t<DbName>biocollections</DbName>\\n\\t<DbName>gtr</DbName>\\n</DbList>\\n\\n</eInfoResult>\\n'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "result"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7cfedad",
   "metadata": {},
   "source": [
    "We can parse this result with the `Entrez` object, returning a python object:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "87e08a1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "stream = Entrez.einfo()\n",
    "record = Entrez.read(stream)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ebc602a",
   "metadata": {},
   "source": [
    "This returns a dictionary with one key, `'DbList'`. This key value is the\n",
    "list of all available databases:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "00b4eaa7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['pubmed', 'protein', 'nuccore', 'ipg', 'nucleotide', 'structure', 'genome', 'annotinfo', 'assembly', 'bioproject', 'biosample', 'blastdbinfo', 'books', 'cdd', 'clinvar', 'gap', 'gapplus', 'grasp', 'dbvar', 'gene', 'gds', 'geoprofiles', 'medgen', 'mesh', 'nlmcatalog', 'omim', 'orgtrack', 'pmc', 'proteinclusters', 'pcassay', 'protfam', 'pccompound', 'pcsubstance', 'seqannot', 'snp', 'sra', 'taxonomy', 'biocollections', 'gtr']"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "record['DbList']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41c62cfa",
   "metadata": {},
   "source": [
    "For each database, we can use `einfo()` again to gain more information:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e67f03cc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['DbName', 'MenuName', 'Description', 'DbBuild', 'Count', 'LastUpdate', 'FieldList', 'LinkList'])"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "stream = Entrez.einfo(db='nucleotide')\n",
    "record = Entrez.read(stream)\n",
    "record['DbInfo'].keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "fe5e8a0b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'688445866'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "record['DbInfo']['Count']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac438e3d",
   "metadata": {},
   "source": [
    "One really useful data from this object return from `einfo(db)` is the 'FieldList',\n",
    "which gives us a list of possible search fields to use with ESearch:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c0375644",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ALL, All Fields, All terms from all searchable fields\n",
       "UID, UID, Unique number assigned to each sequence\n",
       "FILT, Filter, Limits the records\n",
       "WORD, Text Word, Free text associated with record\n",
       "TITL, Title, Words in definition line\n",
       "KYWD, Keyword, Nonstandardized terms provided by submitter\n",
       "AUTH, Author, Author(s) of publication\n",
       "JOUR, Journal, Journal abbreviation of publication\n",
       "VOL, Volume, Volume number of publication\n",
       "ISS, Issue, Issue number of publication\n",
       "PAGE, Page Number, Page number(s) of publication\n",
       "ORGN, Organism, Scientific and common names of organism, and all higher levels of taxonomy\n",
       "ACCN, Accession, Accession number of sequence\n",
       "PACC, Primary Accession, Does not include retired secondary accessions\n",
       "GENE, Gene Name, Name of gene associated with sequence\n",
       "PROT, Protein Name, Name of protein associated with sequence\n",
       "ECNO, EC/RN Number, EC number for enzyme or CAS registry number\n",
       "PDAT, Publication Date, Date sequence added to GenBank\n",
       "MDAT, Modification Date, Date of last update\n",
       "SUBS, Substance Name, CAS chemical name or MEDLINE Substance Name\n",
       "PROP, Properties, Classification by source qualifiers and molecule type\n",
       "SQID, SeqID String, String identifier for sequence\n",
       "GPRJ, BioProject, BioProject\n",
       "SLEN, Sequence Length, Length of sequence\n",
       "FKEY, Feature key, Feature annotated on sequence\n",
       "PORG, Primary Organism, Scientific and common names of primary organism, and all higher levels of taxonomy\n",
       "COMP, Component Accession, Component accessions for an assembly\n",
       "ASSM, Assembly, Assembly\n",
       "DIV, Division, Division\n",
       "STRN, Strain, Strain\n",
       "ISOL, Isolate, Isolate\n",
       "CULT, Cultivar, Cultivar\n",
       "BRD, Breed, Breed\n",
       "BIOS, BioSample, BioSample\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for field in record['DbInfo']['FieldList']:\n",
    "  print(f\"{field['Name']}, {field['FullName']}, {field['Description']}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "122e8812",
   "metadata": {},
   "source": [
    "This is used to restrict our search within a specific field, such as with\n",
    "`Jones[AUTH]` to limit our search to the author field, or `Sanger[AFFL]` to\n",
    "limit out search to author's affiliated to the Sanger Centre."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1e5db5f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "handle = Entrez.esearch(db=\"nucleotide\", term='CRT[Gene Name] AND \"Plasmodium falciparum\"[Organism]')\n",
    "rec_list = Entrez.read(handle)\n",
    "if int(rec_list['RetMax']) < int(rec_list['Count']):\n",
    "    handle = Entrez.esearch(db=\"nucleotide\", term='CRT[Gene Name] AND \"Plasmodium falciparum\"[Organism]', retmax=rec_list['Count'])\n",
    "    rec_list = Entrez.read(handle)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "645900cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "id_list = rec_list['IdList']\n",
    "hdl = Entrez.efetch(db='nucleotide', id=id_list, rettype='gb', retmax=rec_list['Count'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1ef93e81",
   "metadata": {},
   "outputs": [],
   "source": [
    "recs = list(SeqIO.parse(hdl, 'gb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "da74bfc0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "KM288867\n",
       "Plasmodium falciparum clone PF3D7_0709000 chloroquine resistance transporter (CRT) gene, complete cds\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for rec in recs:\n",
    "    if rec.name == 'KM288867':\n",
    "        break\n",
    "print(rec.name)\n",
    "print(rec.description)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "99b9aaa4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "not processed:\n",
       "type: source\n",
       "location: [0:10000](+)\n",
       "qualifiers:\n",
       "    Key: clone, Value: ['PF3D7_0709000']\n",
       "    Key: db_xref, Value: ['taxon:5833']\n",
       "    Key: mol_type, Value: ['genomic DNA']\n",
       "    Key: organism, Value: ['Plasmodium falciparum']\n",
       "\n",
       "['CRT']\n",
       "not processed:\n",
       "type: mRNA\n",
       "location: join{[2751:3543](+), [3720:3989](+), [4168:4341](+), [4513:4646](+), [4799:4871](+), [4994:5070](+), [5166:5249](+), [5376:5427](+), [5564:5621](+), [5769:5862](+), [6055:6100](+), [6247:6302](+), [6471:7598](+)}\n",
       "qualifiers:\n",
       "    Key: gene, Value: ['CRT']\n",
       "    Key: product, Value: ['chloroquine resistance transporter']\n",
       "\n",
       "not processed:\n",
       "type: 5'UTR\n",
       "location: [2751:3452](+)\n",
       "qualifiers:\n",
       "    Key: gene, Value: ['CRT']\n",
       "\n",
       "not processed:\n",
       "type: primer_bind\n",
       "location: [2935:2958](+)\n",
       "qualifiers:\n",
       "\n",
       "not processed:\n",
       "type: primer_bind\n",
       "location: [3094:3121](+)\n",
       "qualifiers:\n",
       "\n",
       "not processed:\n",
       "type: CDS\n",
       "location: join{[3452:3543](+), [3720:3989](+), [4168:4341](+), [4513:4646](+), [4799:4871](+), [4994:5070](+), [5166:5249](+), [5376:5427](+), [5564:5621](+), [5769:5862](+), [6055:6100](+), [6247:6302](+), [6471:6548](+)}\n",
       "qualifiers:\n",
       "    Key: codon_start, Value: ['1']\n",
       "    Key: gene, Value: ['CRT']\n",
       "    Key: product, Value: ['chloroquine resistance transporter']\n",
       "    Key: protein_id, Value: ['AIW62921.1']\n",
       "    Key: translation, Value: ['MKFASKKNNQKNSSKNDERYRELDNLVQEGNGSRLGGGSCLGKCAHVFKLIFKEIKDNIFIYILSIIYLSVCVMNKIFAKRTLNKIGNYSFVTSETHNFICMIMFFIVYSLFGNKKGNSKERHRSFNLQFFAISMLDACSVILAFIGLTRTTGNIQSFVLQLSIPINMFFCFLILRYRYHLYNYLGAVIIVVTIALVEMKLSFETQEENSIIFNLVLISALIPVCFSNMTREIVFKKYKIDILRLNAMVSFFQLFTSCLILPVYTLPFLKQLHLPYNEIWTNIKNGFACLFLGRNTVVENCGLGMAKLCDDCDGAWKTFALFSFFNICDNLITSYIIDKFSTMTYTIVSCIQGPAIAIAYYFKFLAGDVVREPRLLDFVTLFGYLFGSIIYRVGNIILERKKMRNEENEDSEGELTNVDSIITQ']\n",
       "\n",
       "Exon 3452 3543 1\n",
       "Exon 3720 3989 1\n",
       "Exon 4168 4341 1\n",
       "not processed:\n",
       "type: primer_bind\n",
       "location: [4288:4323](-)\n",
       "qualifiers:\n",
       "\n",
       "Exon 4513 4646 1\n",
       "Exon 4799 4871 1\n",
       "Exon 4994 5070 1\n",
       "Exon 5166 5249 1\n",
       "Exon 5376 5427 1\n",
       "Exon 5564 5621 1\n",
       "Exon 5769 5862 1\n",
       "Exon 6055 6100 1\n",
       "Exon 6247 6302 1\n",
       "Exon 6471 6548 1\n",
       "not processed:\n",
       "type: 3'UTR\n",
       "location: [6548:7598](+)\n",
       "qualifiers:\n",
       "    Key: gene, Value: ['CRT']\n",
       "\n",
       "not processed:\n",
       "type: primer_bind\n",
       "location: [7833:7856](-)\n",
       "qualifiers:\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for feature in rec.features:\n",
    "    if feature.type == 'gene':\n",
    "        print(feature.qualifiers['gene'])\n",
    "    elif feature.type == 'exon':\n",
    "        loc = feature.location\n",
    "        print('Exon', loc.start, loc.end, loc.strand)\n",
    "    else:\n",
    "        print('not processed:\\n%s' % feature)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "43b4d147",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "molecule_type=DNA\n",
       "topology=linear\n",
       "data_file_division=INV\n",
       "date=12-NOV-2014\n",
       "accessions=['KM288867']\n",
       "sequence_version=1\n",
       "keywords=['']\n",
       "source=Plasmodium falciparum (malaria parasite P. falciparum)\n",
       "organism=Plasmodium falciparum\n",
       "taxonomy=['Eukaryota', 'Sar', 'Alveolata', 'Apicomplexa', 'Aconoidasida', 'Haemosporida', 'Plasmodiidae', 'Plasmodium', 'Plasmodium (Laverania)']\n",
       "references=[Reference(title='Versatile control of Plasmodium falciparum gene expression with an inducible protein-RNA interaction', ...), Reference(title='Direct Submission', ...)]\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for name, value in rec.annotations.items():\n",
    "  print(f'{name}={value}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "651535ff",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ATATGTAAAACCAAAATAAATTAAACAGAATTTATTTTTAAAAGATTTATTTGTAACAATATTACCATGATGATTTATTAAAGTAAAATCACCACCTATTAATGGTTTTCCTATACTTTCCATTGTAGTTTTTCCAAAACCTTTTTTTTTATTTTTCTCATTTTGTAATAAATATAAATATAAATATAATGCTGGTATCGTTAATGATAAATTAATACCTAAACATTTCCATGTTAAAAATATATTTTTCTTTTTCGACTTTTTTTTATTATCATTTTCATCCTCAACTTTTTTACTCATACTACTATAATCATTTCTTATAAAACTTTTATAAAATATATTTTCCCTTTTTTTATAATTTCTACTAGTATCATAAACAATATTTCTTATATTCAATAATACATTAAAATTGTTATATCCATTTTTTTTTTTATTACATATTTTGTTTAAGCTATTTAAAAGCACCTTATTCATTTTTGAAATAATGAAACAGCAAAAAAACAAAAATAACGTAAACAATTATGATGTACATAAAATATATATATATATATATATATATATATATTATTTATATATTATTCACATTTACATTTATTTATTTATATCTTCCTATTTTATACGATAAAATAAATTCGATTTAAATAAAATATACGTTAAATTAAATTCTTATATCTTAACATTTCGACATACATTTATTAATTGATTTTATTTTATTTATTTTGTTATATCCATATAAATAAACGTATCGCACATGAGTTTTTTTTTTTTTTTTTTTTTTTTTTTATATAATACTTTTTAATAGTGAAACACATACATATTTTTTACCATTTTAAATATATATATGGTGTACCATTTTGGGTTAATATATTTTTTGTAATTATATATTAATATATACATAATAAATATAAATTAATATTATATATATATATATATATATATATATAAATAATTGGAGTGTCCTATTTTGAAGCTATGTCCTTTGGAATACAGCACTTTACAAAAAAAAAATTATTTTATAATAAATGGAAAAAAAAGAACAGAACAAAATAATACATATTTATATTTAATACCAAAAAAAAAAGAAAAAAAAAAGAAAAAGAAAAAAAAAAAACTCCTATATATATATTATAAACGAATTACATCCTAATTTTTTTTTTTTTCTTCTTTTTAAATATAATAATTATTTATATGCCATATTTATTTATTATACATATATATATATATATATATATATATATATTACATTAAAAAATTAAAAAAAAGTAATATATTAAATAATATGTTCGTATTTTTTTTATTTATGTTTGTATTTTTATTTTTTTGTATAATCTATGAAAAGAAATAATTACCTATATAATATATTATATAATATATATATATATATATAGTCTAAATAAAAAAATACATATAAAATATATATACAGTATCAAAAAAATGTAATGTATAAAAAATAAAATATTGGGGGTAGAGAAATATTTTTATAAGGGTTAAATAATAGGCCTATTTTGTTTTTATATACATTATGGCTTAAAAGAAGAAATATATATATATATATATAATTATAATTATATATAAAATTTTTTTTAAGGTTTTAAATAAATACATATTTATTTATGTATTAATATATTATATATATATATATATATATATATATTATATTACCACATTACCATAATATTAAATTATATATATATTATATATATATATATATATATATATATATTATATATATATATATATATATTAATTATATATATATAAATATTACCACATCATCATAATATTAAATTTATATATATATATATATATATATATTATAATAATAAATATAATTTTATACTTCAATTCTCATATTTCAATATATTCCTTTTAAAGCTCTAAATAGATTCTAAAAAATTAGAACCCTAAGAATATCAGCTCATATATATATGTAAATAAAAATAAAAGTAACTCCTACTGTATTTATATAATATTTAGAAATTTTATATAGTAGTTGAGTGATTCTATATACATATACATAAATAAATTACATAATATTATAAATTTTTTTTATATTATATTAGAATGTTTACTATATAAAAATAAATATTTTCTTATATATTTTTTTATTTTTCCATATGAAAAAAGAAATTTTTTATATATTTTTTTATTATATTTTATAAATAAAAAAGAACTTATACCTATTATTATTATTATATATAAAAAATATATATTTTTTATAAATTCCTATTTTTCCGATTATTTATTTATTTTTTTTTTTTATTGTTTAAAAATATATAAAAAAAATTCTTATATATTTAATATTATAAGTCCATAAATATATATAATTTATATAATTTATATATTTCCATACTTTTTTTATTATATATCATGAGATAAAATAAATTTCATATAGAAATCGTATTTATTTATTAATTTGATATATATTAATAAATAAATATTAATAATAATATGTTATATATATATATATATTTATTTAATTATTATATAGGAAACTATATATATATATATATTATATTTTTTTTTTTTTGAAATATAATAATATATAATATTATTCTAAAAATATCTATATTCTTATACTTGAACCTTTTTTTTTTTTTTTTTTTTTTTTTTTGACTTTCGATTGTTCATTCTGTTTATTGTATAAATATATAAATATATATATATTGTATATTTTATTACATATATTTTTTTTTTTTGAAAGTTAAGAATTTTAACTTAATAAAAAAAGTACATATATTTTTAATAAATGTCCTCCATTATATAAATTGTTATATAAAAGATTTTATATAATTTAAATAGAATTCATTTATAAACAAATTTGTTTATAAAAATATATTTATGTATATATAATATAAATATATATATTTATATATATATATATATATATATATTACTATATATATTTTTTTTTTTTTTTCCTTTTTTTTACTTTCCCAAGTTGTACTGCTTCTAAGCTTTTTTTAATAAACATATATATAATTTGTACAAATATTTTAGATTATATACATGATGTATATTTGAATATATTTTCTATATATTTGTGGTTCCATTTTTGTATATTATATATAATATATTTATATATATATTGATATGTCAATATTTGTATAACACATGAAGTTTTTGTTTTTTTTTTTTTTTTTTTTTTAATGGAGAATATTTAATAATATATGAAAAAAATTTTATATAAATATATATATATATATATATATATATATATATATATATGTATATATATTTATATATACATGTATGTTTTTTAAAAAGTTAAATAATTCTATAGATTATTTTCATTGTCTTCCACATATATGACATAAATATTTTAAAATCGACATTCCGATATATTATATTTTTAGACTATAATATCCGTTAATAATAAATACACGCAGTCATATTATTTATTATACATTCATTTATTATTTTGTTTTTTTTAATTTCTTACATATAACAAAATGAAATTCGCAAGTAAAAAAAATAATCAAAAAAATTCAAGCAAAAATGACGAGCGTTATAGAGAATTAGATAATTTAGTACAAGAAGGAAGTAAGTATCCAAAAATGGAAATATGGAATGATATAAATGAATAGATAAATCAACCTATTGGATATATATATATATATATATATATATATATATATATATATGTATACCCATATGTATTAATTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTCCCTTGTCGACCTTAACAGATGGCTCACGTTTAGGTGGAGGTTCTTGTCTTGGTAAATGTGCTCATGTGTTTAAACTTATTTTTAAAGAGATTAAGGATAATATTTTTATTTATATTTTAAGTATTATTTATTTAAGTGTATGTGTAATGAATAAAATTTTTGCTAAAAGAACTTTAAACAAAATTGGTAACTATAGTTTTGTAACATCCGAAACTCACAACTTTATTTGTATGATTATGTTCTTTATTGTTTATTCCTTATTTGGAAATAAAAAGGGAAATTCAAAAGTAAGATAAATCAATATATTAAAATGATGGATTTATAAGAGAATCTATTCCACCTACCAATATAAAACATTACACATATATATATATATATATATATATATATATATATATATATATATATGTATGTATGTTGATTAATTTGTTTATATATTTATATTTATTTCTTATGACCTTTTTAGGAACGACACCGAAGCTTTAATTTACAATTTTTTGCTATATCCATGTTAGATGCCTGTTCAGTCATTTTGGCCTTCATAGGTCTTACAAGAACTACTGGAAATATCCAATCATTTGTTCTTCAATTAAGTATTCCTATTAATATGTTCTTCTGCTTTTTAATATTAAGATATAGGTAAGTATACTATTTTAAATTACAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAATATAAAATATATATATATATTTATATATATTTATTTATATATTTATATATATTTATTTATATATTTATATATTTATTTATTTATATTTATTTATTTACTCCTTTTTAGATATCACTTATACAATTATCTCGGAGCAGTTATTATTGTTGTAACAATAGCTCTTGTAGAAATGAAATTATCTTTTGAAACACAAGAAGAAAATTCTATCATATTTAATCTTGTCTTAATTAGTGCCTTAATTGTAAGAAAACAAAATATATAAATAAATAAATATATATATATATATATATATATTGTAATATTTAATATATATTAACACTTAAACTTTGTTTTTATTATATTAATTTATATTCTTTATCATTTTTTTTTTTTTTTTTCCTTCTTTTTTTTTTAGCCTGTATGCTTTTCAAACATGACAAGGGAAATAGTTTTTAAAAAATATAAGATTGACATTTTAAGATTAAATGTAAGAAGAAAAATATAATAATAATAATATATATATTATATATATATTTCCTTTTTTACCACTTTTTTTTTTTTTATTCCTATAACGCATTATAATTATTTCTGTTATTTTTATTTCTTATAGGCTATGGTATCCTTTTTCCAATTGTTCACTTCTTGTCTTATATTACCTGTATACACCCTTCCATTTTTAAAACAACGTAAGAATTAATTAGGAAAGAAAAAATAAAATAAAATGAAATGTGCCCATATATATATATATATATATGTATGTATGTATAATTTTCCCTTTTTAGTTCATTTACCATATAATGAAATATGGACAAATATAAAAAATGGTTTCGCATGTTTATTCTTGGGAAGAAACACAGTCGTAGAGGTAAAATAGGATTTTCATTATATATTAAAAATTACTACTTATTATGTTAATAAAAAAATATGTTTTTTAATGTTCAATTTGTTTTATTTTAATTATTTTTTTTTTTTTTGTTTTGTTTCCTCTTCAGAATTGTGGTCTTGGTATGGCTAAGTTATGTGATGATTGTGACGGAGCATGGGTAAGAAGCTTATAATAAAATTTCAAAATTATAAGAGACATTTATATATATTTTAACAATAATAATTAAATAAAACAATATTATATATATTATATATATTATTATTTTATTTATTATTTTTTATTTTTTTTTTTTAGAAAACCTTCGCATTGTTTTCCTTCTTTAACATTTGTGATAATTTAATAACCAGCTATGTAAGAATAAAAAGGATGAAAAACGTTGATACATACAAGAGGAAATAAATATGGGAATGTTTAATTGAATTAAGATATATATATATATATATATATATATATATATATATGTAACCATATAATTTTTCATTTTCTATGCTTTTTATAGATTATCGACAAATTTTCTACCATGACATATACTATTGTTAGTTGTATACAAGGTCCAGCAATAGCAATTGCTTATTACTTTAAATTCTTAGCCGTAAGAATTAAAAAGATATAAATATATAAATATATATGTGAAATATATGATATATATGATATATATTTTTTATATGTAATGTTTTTTTTTTTTTTCACAATATACATTTAAATGTTTATGATGGTACAACGTATCATATTTTATAATAATTTTATGCATTCATGTATATTATTTTTACTTTTTAATTTTATAGGGTGATGTTGTAAGAGAACCAAGATTATTAGATTTCGTAACTTTGGTAAGTGTGAAATTAAAAAATGAATTTTTTTTTTTTTTTTTTATGAACAAAATAATGTGTATATAATATGTGTGTAATATCAAATGGCTTGTTCGTTCATAAATATTTATATTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTACAGTTTGGCTACCTATTTGGTTCTATAATTTACCGTGTAGGAAATATTATCTTAGAAAGTAATACAAAAATAAGATAAAAAAATATAATATATAAAATATGTATATTGTTCTTATATATTTTGTTCATATATATATATATATATATATATATTTTTATATTTCCATCTGTCTTTTTATTCTATTGTTATAATTTATCATAAATTTTTTTTAAATTTGCTTACATTAGGAAAAAAAATGAGAAATGAAGAAAATGAAGATTCCGAAGGAGAATTAACCAACGTCGATTCAATTATTACACAATAAAATTATATCATTTATTTTTTTATATTTTTTATATTTTTTTTATTTTTCATGTTTTTTTCTTTATTTTTTTTTTTTTTTTTTTCCTTTTTTTTTTTAATAAATCATTTTCCTAATATTTATATTTGCTATCGCATTACACTTTATAAGGAATAATGTTGAAGGTATAGAGATCTCTTTATTATTATTAAGATGTAAATTTTTTATGAAATTTTGTACTTTTTTAGTATTTATATTATATACACATAAATATATATATATTTATTTATTGATATTTGAAAAAAGAATAAGGAAATAAAAAAAGGGAAAAAAAAAAAAGTTAAAAGTATATATATATATATATATATATATATTTATTTATTTATATATATTATTATGTATGTTTATAATTTATATATATATGTCATAAAAAGAAATATAATTATATATATATATATTATATTATATATATATATAAATTTTTTTAAATTATAAACTATTGTTATTGTTAAATTTAATTTTTAATTCATTAAAAATCAGATTACCAAAGATTTTTTCCATTCTTCAATCATGATAATAGAAATGAGAAGAAGCAAAAAAAGAAAAAAAAAAAAAATTCATACTGTGTCAAAGGAGCAAACAATTTTGTCCAATTGACAAATATGTATATATATAAAATATATATATATATATATATATATATATATATATATATATATATTTATTGATTTATTCCTTTTCTTTTTTTTTTCCCTTTTGGTGTTTTGATAAATAGGTAAAAAATAAAATATATAAAAATTATCTCTAACATTCGTAAAGATTTATTTTATTTTTTAATTTTTTCTATTTTATCAAGTTTGTAAATTATCATTTCATTAATGAAATAATAAGAATACAAAATGTTTTGTATTTGTTCTTGTTAAGAAGATTTTACAATTATATTTTTTTTTTTAATTTATTTTATAAAATATAATTAAGATTATATCTTCTAAAAGAAAAAAAAAAAAGAAAAAAGAAATATTTTGTATAATGAAATTTGAAATATCTTTCGCCACTTATTTCAAGTTTCAGTTATTTTATTTTATTTTATTTTTTTTTTTTTTTTCTTGCTTCTTCTCTTTCTTCTTGTATAATATTAGTTATCATCATCTTTTTAACGAATTAAATACAAAATAAATGTATTTTACTAAAAAAAGAAAAAAAAAAAAAAAAGTATGTTATATTTTTTTTTTTTTGTTAAATAATGCCTTATTCAAAGATAATGTTTTTTTAATAAAACAGAAATATTATTATATTTATTCTTATAAAATGGGTGGAATATATAAGGGTCAAAAAAAAAAAAAATAATAAAATAAATAAATAAATAAATAAAATATATAAATAAAATAATAATAAAATAAAAGGAATAAATAATTATATAATATATATATTATATATATATATTATGTATATATATTTTATATATTCCAGTATGTTCGCGATAAAAAAAAAAAAATATATTTAATAAATAAAAATTACAAGGACCAGAAAAATATATACTGCACATATATATATGTATATATATATATATATATATAATTATTTAATATAATAATTTTTTTGTAAATCCCATTGTATCATTAATATTGAAATTTAATTTCTTATTTCTTATTTTTTTTATATTACACATTCGTACATACATAATATATATATATATATATATAAGAGAAAATAATAAAACACGTTAATATATATATATATATATATATATATATTATATATATATATTTATTTATTTATTTATATATTATTACGCATTAATCATTTAAAGTACCATTTTGTATTTTTTTTTTTTTTTTTGAAGTTATATTATATATATATATATATATATAAAATATATCGAATAATGAAAAATATCAGAGTAAATATGATATAAAAGAAGATGTTTACTTTTCTTGTTATTATTGTAAACGTATTGGTTTGCATATGTTTCTTAAATACAAAATACCAGATTCAACTGGATTATAGTGTACCTAGTACCTTAATCAATTTAGATAGATCTTTAAAATATACGAACATAACGGTAGGAAAAGATTCCTTTCTTTGTATATTAGATAAGTCTGATAGTTTTGATACATGTGAGCAATATGATTTATTGAAAGAGGTTGATGATTTTTACCCTAATGGGTTTGACGAACATAAAGAAATAAATAATAATAATAATAATAATAAAAAGAAGAAGAGTTATAATAATAATTATAAAAATAATTATAATCATATCAGCAATAATGGTGACAATCAAATTAATAATGATTATGATACAACCTACAATAGGTATGATTATATAGAAGAACATCTAAATAAAAACTTTCTTCTAAAATATCAAGATAATAATAATATGAACTATAATTTATACCTAAAGGAAGATAACTACGAAATCGAATATCATAATGCACATTACTGGATGAATGTCCAAGATGAAAATATACAAGCAAGCGATCTTACACAAAATATTATTAAAGATGTTGGTGATAATAATACAAGTTATAAGAACAGTAATAATAACGGTTGGACTTGTAATAACGTATTGAATAAATTAAATAATAAATATTCCAAAGAATTAAAGGGGGGGAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAGATAAAAAAAAAAAAAATCATTATGATTATTTTTCGAACGAAAGAAACGAAAAAGTTGATAAATTAAATAAGACATTTAATTGTAATTTTAAAAGGTTATATGATTATATTTTTTATAAACATAAAATTAATGCGTATCATAAATATAGAATATATAATAATATTAATGATGACGTATTATTAAAAAGTCATATAATAAAAGGGAAAATGTTAACCATAAATAATACATGTATTGATGCGTTTAATAATGGAAATATATTAAAAATATGTTTAAATAAATCTATATCATTTATAACAAAAAGATATCATACGAAACATAAAAAACATATCACTATATCTACTTATGCTATAGGAAAAGATTTATTATTTTCTAATAATACTGTTGTACAATATTATAGTGATGGCAAATATTTTTTTGAAGTATTATTTTTATGTGGTAATAATAATTTAAGAGTTAATAGTTTTGAAAAGCTGTATAATGTATATTATCCATTCATTTTCCAAATATATGAACATTTAAATACAACTATTACTAAACTTTATAATATTATAATAAAAAAATTAAAAGATAATGATGTATATTTTTTTAAATCATTTTCGAATTATCAATATACTGAAGATTATTATAAATCATTAAAACAAAAATTAATACACAATTTAAATTTATATTATCCTGAAGCTATATATTTATATCGTATTACATTAAGTGGATATATGTTTTGTAATTATACAGAGAAAATAAATCCACCAAATTTTTTTCTTTTAAAAAATCTGATGAATAAATGTCATCATTTTATTAGAGAAGATGATACCATGTTTGAAATATGTTTACCAT\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "print(rec.seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4b294b9",
   "metadata": {},
   "source": [
    "## Performing basic sequence analysis\n",
    "\n",
    "We will be using the human **lactase** gene as an example. We will use the\n",
    "same methods as above for retrieving it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4de4173e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez, SeqIO, SeqRecord\n",
    "Entrez.email=\"guilhormo.47@gmail.com\"\n",
    "hdl = Entrez.efetch(db='nucleotide', id=['NM_002299'], rettype='gb') # lactase gene\n",
    "gb_rec = SeqIO.read(hdl, 'gb')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "675f3f9a",
   "metadata": {},
   "source": [
    "Now that we have the record, we'll extract the gene sequence, even though it\n",
    "contains more than that."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "334ab4cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "for feature in gb_rec.features:\n",
    "  if feature.type == 'CDS':\n",
    "    location = feature.location # Note translation existing\n",
    "cds = SeqRecord.SeqRecord(gb_rec.seq[location.start:location.end], 'NM_002299', description='LCT CDS only')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e91c185",
   "metadata": {},
   "source": [
    "Now our sequence is available in the Biopython sequence record.\n",
    "\n",
    "We'll now proceed by saving the sequence to a FASTA file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e6bcb94",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "with open('example.fasta', 'w') as w_hdl:\n",
    "  SeqIO.write([cds], w_hdl, 'fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec86a0dc",
   "metadata": {},
   "source": [
    "The Seq.IO write method takes a list of sequences to write. In our case it is just one,\n",
    "but we should be careful - if we plan to write lots of sequences, an iterator\n",
    "will allocate less memory from our computer.\n",
    "\n",
    "In most situations we will have the sequence loaded on the disk and will\n",
    "want to read it. For that, we have the `SeqIO.read()` method:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f15acbbc",
   "metadata": {},
   "outputs": [],
   "source": [
    "recs = SeqIO.parse('example.fasta', 'fasta')\n",
    "for rec in recs:\n",
    "  seq = rec.seq\n",
    "  print(rec.description)\n",
    "  print(seq[:10])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71cb9065",
   "metadata": {},
   "source": [
    "As we now have pure DNA, we can transcribe it:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cab269b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna = seq.transcribe()\n",
    "print(rna)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee0c14b1",
   "metadata": {},
   "source": [
    "And then, we can translate the gene into a protein:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54baca52",
   "metadata": {},
   "outputs": [],
   "source": [
    "prot = seq.translate()\n",
    "print(prot)"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "formats": "ipynb,md"
  },
  "kernelspec": {
   "display_name": "bio_genbank",
   "language": "python",
   "name": "bio_genbank"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
